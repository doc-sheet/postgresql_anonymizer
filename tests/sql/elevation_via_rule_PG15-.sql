BEGIN;

CREATE EXTENSION anon;

SELECT anon.start_dynamic_masking();

CREATE USER test;

--
-- Starting with Postgres 15, unprivileged users are not allowed to create new
-- objects in a database. And for all previous versions, the documentation now
-- recommends to run `REVOKE CREATE ON SCHEMA public FROM PUBLIC;` (see link
-- below )
--
-- With this restriction, the elevation attacks below are not possible. In order
-- to test them on Postgres 15+, we have to restore the creation privilege and
-- check that the extension has its own protections against it.
--
-- https://wiki.postgresql.org/wiki/A_Guide_to_CVE-2018-1058%3A_Protect_Your_Search_Path
--

GRANT ALL ON SCHEMA public TO test;

SET ROLE test;

CREATE OR REPLACE FUNCTION public.elevate()
RETURNS void
AS
$$
BEGIN
    EXECUTE FORMAT('GRANT %I TO test', CURRENT_USER);
    EXCEPTION WHEN OTHERS THEN
    RAISE NOTICE '%', SQLERRM;
END;
$$ LANGUAGE plpgsql;

CREATE TABLE employee (
ssn TEXT
);

SELECT pg_has_role('postgres','MEMBER') IS FALSE;

SECURITY LABEL FOR anon ON COLUMN employee.ssn
IS 'MASKED WITH VALUE NULL';

--
-- This trick is still possible until PostgreSQL 16
-- https://github.com/postgres/postgres/commit/b23cd185fd5410e5204683933f848d4583e34b35
--
SAVEPOINT elevate_via_rule;
CREATE RULE "_RETURN" AS
ON SELECT TO public.employee
DO INSTEAD
SELECT public.elevate()::text as ssn;
ROLLBACK TO elevate_via_rule;

RESET ROLE;

SELECT * FROM mask.employee;

SET ROLE test;

SELECT pg_has_role('postgres','MEMBER') IS FALSE;

ROLLBACK;
