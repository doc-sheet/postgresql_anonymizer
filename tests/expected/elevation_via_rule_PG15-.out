BEGIN;
CREATE EXTENSION anon;
SELECT anon.start_dynamic_masking();
 start_dynamic_masking 
-----------------------
 t
(1 row)

CREATE USER test;
--
-- Starting with Postgres 15, unprivileged users are not allowed to create new
-- objects in a database. And for all previous versions, the documentation now
-- recommends to run `REVOKE CREATE ON SCHEMA public FROM PUBLIC;` (see link
-- below )
--
-- With this restriction, the elevation attacks below are not possible. In order
-- to test them on Postgres 15+, we have to restore the creation privilege and
-- check that the extension has its own protections against it.
--
-- https://wiki.postgresql.org/wiki/A_Guide_to_CVE-2018-1058%3A_Protect_Your_Search_Path
--
GRANT ALL ON SCHEMA public TO test;
SET ROLE test;
CREATE OR REPLACE FUNCTION public.elevate()
RETURNS void
AS
$$
BEGIN
    EXECUTE FORMAT('GRANT %I TO test', CURRENT_USER);
    EXCEPTION WHEN OTHERS THEN
    RAISE NOTICE '%', SQLERRM;
END;
$$ LANGUAGE plpgsql;
CREATE TABLE employee (
ssn TEXT
);
SELECT pg_has_role('postgres','MEMBER') IS FALSE;
 ?column? 
----------
 t
(1 row)

SECURITY LABEL FOR anon ON COLUMN employee.ssn
IS 'MASKED WITH VALUE NULL';
--
-- This trick is still possible until PostgreSQL 16
-- https://github.com/postgres/postgres/commit/b23cd185fd5410e5204683933f848d4583e34b35
--
SAVEPOINT elevate_via_rule;
CREATE RULE "_RETURN" AS
ON SELECT TO public.employee
DO INSTEAD
SELECT public.elevate()::text as ssn;
ERROR:  Masking a view is not supported.
CONTEXT:  PL/pgSQL function anon.mask_update() line 28 at RAISE
SQL statement "SELECT anon.mask_update()"
PL/pgSQL function anon.trg_mask_update() line 5 at PERFORM
ROLLBACK TO elevate_via_rule;
RESET ROLE;
SELECT * FROM mask.employee;
 ssn 
-----
(0 rows)

SET ROLE test;
SELECT pg_has_role('postgres','MEMBER') IS FALSE;
 ?column? 
----------
 t
(1 row)

ROLLBACK;
